### ISSUE TITLE
How to dynamically infer the number of nodes in GraphsTuple

### ISSUE THREAD

**Issue by ferreirafabio** [deepmind/graph_nets#61]
Hi, in the `EncodeProcessDecode` model, I am trying to reset the node state in every rollout step in the for loop of the `_build` function to the ground truth node state. The networkx graphs that I used to create the create the `target_op` (target_graphs) and `input_op` (input_graph) are:
```
input_graph: (1,)
target_graphs: (14,)
```

`num_processing_steps` is 14 and the code I am using is the following:

 ```
        ground_truth_graphs_T = self._encoder._network._node_model(target_op.nodes)
        n_nodes = [tf.shape(ground_truth_graphs_T)[0]]
        mult = tf.constant([num_processing_steps])
        ground_truth_graphs_splits = tf.split(ground_truth_graphs_T, num_or_size_splits=tf.tile(n_nodes, mult), axis=0)


        for step in range(num_processing_steps):
            ground_truth_graphs_t = ground_truth_graphs_splits[step]
            latent = latent.replace(nodes=ground_truth_graphs_t)

            latent = self._core(latent)
            decoded_op = self._decoder(latent, is_training)
            output_ops.append(decoded_op)
```
I apply `tf.split` because I assume `target_op.nodes` has shape `(?, x)` while `?` being `n_node*num_processing_steps`.
This code fails with the following error and I fail to understand where the problem is:

> InvalidArgumentError (see above for traceback): Determined shape must either match input shape along split_dim exactly if fully specified, or be less than the size of the input along split_dim if not fully specified.  Got: 196
> 	 [[node EncodeProcessDecode/split (defined at /scr2/repos/GN/models/EncodeProcessDecode_one_step.py:134)  = SplitV[T=DT_FLOAT, Tlen=DT_INT32, num_split=14, _device="/job:localhost/replica:0/task:0/device:GPU:0"](_arg_placeholders_from_networkxs_1/nodes_0_11/_63, EncodeProcessDecode/Tile/_65, gradients/EncodeProcessDecode/node_model_13/visual_and_latent_node_encoder/concat_grad/concat_1/axis)]]
> 	 [[{{node EncodeProcessDecode/MLPDecoderGraphIndependent_13/graph_independent_1/global_model/mlp_decoder_global/sequential/linear/add/_6037}} = 
>       _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_11958...linear/add", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

## Question
I assume the number of nodes is not correctly inferred. How do I get it from GraphsTuple `input_op` or `target_op` after graph definition time? In the tutorial I found this piece of information:

`nodes (shape=[total_num_nodes] + node_feature_dimensions)`
--> What does the "+" here mean? Assume a graph with 3 total nodes and a node feature dimension of 200. What will the shape look like? Thank you.

**Comment by ferreirafabio** [comment]
@alvarosg you seem to be one of the authors of the new tutorial. It would be great if you could clarify this. Thank you

**Comment by alvarosg** [comment]
Hi Fabio,

I am not sure what you mean when you say:
input_graph: (1,)
target_graphs: (14,)

Do you mean lists of graphs tuples?

Also, when you say:
"I apply tf.split because I assume target_op.nodes has shape (?, x) while ? being n_node*num_processing_steps."

I am not sure why you expect that and it is hard to say without seeing the code that you use to generate your inputs.

With respect to the two specific questions:

* You can get the total number of nodes in a batch of graphs using:

`tf.reduce_sum(graphs_tuple.n_node)`

* The "+" there is simply concatenating the 
shape. Assuming there are `total_num_nodes=145` in the batch, and each node has unidimensional features (the most common case) of size `node_feature_dimensions=[27]`, then the shape of nodes would be
[145, 27].

Hope this helps

**Comment by vbapst** [comment]
The number of nodes in a graph will be given by

`graph.n_node[i]`

where `i` is the index of the graph that you care about.

Note that it looks like your output has size 14 x 14 instead of 14 x 1. What is your batch size?

**Comment by ferreirafabio** [comment]
Sorry, I should've be more clear about this. Yes, correct:

```
input_graph: (1,)
target_graphs: (14,)
```

means 2 lists, one containing one GraphsTuple and one containing 14 GraphsTuple. 


> Also, when you say:
> "I apply tf.split because I assume target_op.nodes has shape (?, x) while ? being n_node*num_processing_steps."
> 
> I am not sure why you expect that and it is hard to say without seeing the code that you use to generate your inputs.

I created both `input_graph` and `target_graphs` as follows:
```
# generate_graph returns a networkx DiGraph 
# with 3 nodes, edges and a feature size of 134406:
graph = generate_graph(config, n_nodes)

# use 'graph' as a template to create two lists 
# (input_graphs with one networkx graph and target_graphs with 14 networkx graphs)
input_graphs, target_graphs = graph_to_input_and_targets(config, graph, batch_data[i])
```
next, I create the ops:
```
    input_op = utils_tf.placeholders_from_networkxs(input_graphs, force_dynamic_num_graphs=True)
    target_op = utils_tf.placeholders_from_networkxs(target_graphs, force_dynamic_num_graphs=True)
```
I then run the `EncodeProcessDecode` model which iterates over these ops (see initial post). If I do `data = sess.run("target_op": target_op, ...)` and evaluate `data["target_op"]`, I get one GraphsTuple object and `target_op.nodes` has shape `(42, 134406)` which the first dimension should exactly be given by `n_node*num_processing_steps = 3 * 14`. Does this sound reasonable?


> Note that it looks like your output has size 14 x 14 instead of 14 x 1. What is your batch size?
Not sure if this has been addressed by the above now. I've been using a batch size of 1

**Comment by vbapst** [comment]
Sorry, so what is your issue exactly?

**Comment by ferreirafabio** [comment]
Hope your questions have been answered. My question now would be: considering your last questions, does something strike you odd? 

I will try out the suggestions for inferring the number of nodes and will report back asap! Thanks!

**Comment by alvarosg** [comment]
Closing due to inactivity. Please feel free to reopen if necessary.
